File: level3/가장_먼_노드.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>
#include <algorithm>

using namespace std;

int visited[20001];

int solution(int n, vector<vector<int>> edge) {
    int answer = 0;
    queue<int> next;
    bool edges[n+1][n+1];

    for(int i=1; i<=n; i++){
        visited[i] = 0;
        for(int j=1; j<=n; j++)
            edges[i][j] = false;
    }

    visited[1] = 1;
    next.push(1);

    for(int i=0; i<edge.size(); i++){
        edges[edge[i][0]][edge[i][1]] = true;
        edges[edge[i][1]][edge[i][0]] = true;
    }

    while(!next.empty()){
        for(int i=1; i<=n; i++){
            if(edges[next.front()][i] && !visited[i]){
                next.push(i);
                visited[i] = visited[next.front()] + 1;
            }
        }
        next.pop();
    }

    int max = 0;
    for(int i=1; i<=n; i++){
        if(max == visited[i])
            answer++;
        else if(max < visited[i]){
            max = visited[i];
            answer = 1;
        }
    }
    return answer;
}

File: JavaScript/level3/가장_먼_노드.js
function solution(n, edge) {
    let graph = Array(n).fill(null).map(()=>Array());
    let visited = Array(n).fill(-1);
    let queue = [0];
    
    visited[0] = 0;

    for(let i of edge){
        graph[i[0]-1].push(i[1]-1);
        graph[i[1]-1].push(i[0]-1);
    }

    while(visited.includes(-1) && queue.length>0){
        let val = queue.shift();
        for(let i=0; i<n; i++){
            if(visited[i] === -1 && graph[val].includes(i)){
                visited[i] = visited[val] + 1;
                queue.push(i);
            }
        }
    }
  
    let val = Math.max(...visited);
    return visited.filter((e) =>  val === e).length;
}